cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3431 @@
+      RETURN_NOT_OK(P_PARTKEY(thread_index));
+      const int32_t* p_partkey = reinterpret_cast<const int32_t*>(
+          tld.part[PART::P_PARTKEY].array()->buffers[1]->data());
+
+      size_t ibatch = 0;
+      int64_t ipartsupp = 0;
+      int64_t ipart = 0;
+      int64_t ps_to_generate = kPartSuppRowsPerPart * tld.part_to_generate;
+      for (int64_t irow = 0; irow < ps_to_generate; ibatch++) {
+        RETURN_NOT_OK(AllocatePartSuppBatch(thread_index, ibatch, PARTSUPP::PS_PARTKEY));
+        int32_t* ps_partkey =
+            reinterpret_cast<int32_t*>(tld.partsupp[ibatch][PARTSUPP::PS_PARTKEY]
+                                           .array()
+                                           ->buffers[1]
+                                           ->mutable_data());
+        int64_t next_run = std::min(batch_size_, ps_to_generate - irow);
+
+        int64_t batch_offset = 0;
+        for (int64_t irun = 0; irun < next_run;) {
+          for (; ipartsupp < kPartSuppRowsPerPart && irun < next_run; ipartsupp++, irun++)
+            ps_partkey[batch_offset++] = p_partkey[ipart];
+
+          if (ipartsupp == kPartSuppRowsPerPart) {
+            ipartsupp = 0;
+            ipart++;
+          }
+        }
+        irow += next_run;
+        tld.partsupp[ibatch][PARTSUPP::PS_PARTKEY].array()->length = batch_offset;
+      }
+    }
+    return Status::OK();
+  }
+
+  Status PS_SUPPKEY(size_t thread_index) {
+    ThreadLocalData& tld = thread_local_data_[thread_index];
+    if (!tld.generated_partsupp[PARTSUPP::PS_SUPPKEY]) {
+      tld.generated_partsupp[PARTSUPP::PS_SUPPKEY] = true;
+      RETURN_NOT_OK(P_PARTKEY(thread_index));
+      const int32_t* p_partkey = reinterpret_cast<const int32_t*>(
+          tld.part[PART::P_PARTKEY].array()->buffers[1]->data());
+
+      size_t ibatch = 0;
+      int64_t ipartsupp = 0;
+      int64_t ipart = 0;
+      int64_t ps_to_generate = kPartSuppRowsPerPart * tld.part_to_generate;
+      const int32_t S = static_cast<int32_t>(scale_factor_ * 10000);
+      for (int64_t irow = 0; irow < ps_to_generate; ibatch++) {
+        RETURN_NOT_OK(AllocatePartSuppBatch(thread_index, ibatch, PARTSUPP::PS_SUPPKEY));
+        int32_t* ps_suppkey =
+            reinterpret_cast<int32_t*>(tld.partsupp[ibatch][PARTSUPP::PS_SUPPKEY]
+                                           .array()
+                                           ->buffers[1]
+                                           ->mutable_data());
+        int64_t next_run = std::min(batch_size_, ps_to_generate - irow);
+
+        int64_t batch_offset = 0;
+        for (int64_t irun = 0; irun < next_run;) {
+          for (; ipartsupp < kPartSuppRowsPerPart && irun < next_run;
+               ipartsupp++, irun++) {
+            int32_t supplier = static_cast<int32_t>(ipartsupp);
+            int32_t partkey = p_partkey[ipart];
+            ps_suppkey[batch_offset++] =
+                (partkey + (supplier * ((S / 4) + (partkey - 1) / S))) % S + 1;
```

Comment:
What is this? Some magic formula from TPC-H? Can you please add a comment?

Reply:
yes, it is magic formula.